test/helpers/envtest.go

```diff
@@ -308,21 +310,88 @@ func (t *TestEnvironment) Cleanup(ctx context.Context, objs ...client.Object) er
 	for _, o := range objs {
 		err := t.Client.Delete(ctx, o)
 		if apierrors.IsNotFound(err) {
-			// If the object is not found, it must've been garbage collected
-			// already. For example, if we delete namespace first and then
-			// objects within it.
 			continue
 		}
 		errs = append(errs, err)
 	}
 	return kerrors.NewAggregate(errs)
 }
 
+// CleanupAndWait deletes all the given objects and waits for the cache to be updated accordingly.
+//
+// NOTE: Waiting for the cache to be updated helps in preventing test flakes due to the cache sync delays.
+func (t *TestEnvironment) CleanupAndWait(ctx context.Context, objs ...client.Object) error {
+	if err := t.Cleanup(ctx, objs...); err != nil {
+		return err
+	}
+
+	// Makes sure the cache is updated with the deleted object
+	errs := []error{}
+	for _, o := range objs {
+		// Ignoring namespaces because in testenv the namespace cleaner is not running.
+		if o.GetObjectKind().GroupVersionKind().GroupKind() == corev1.SchemeGroupVersion.WithKind("Namespace").GroupKind() {
+			continue
+		}
+
+		oCopy := o.DeepCopyObject().(client.Object)
+		err := wait.ExponentialBackoff(
+			wait.Backoff{
+				Duration: 100 * time.Millisecond,
+				Factor:   1.5,
+				Steps:    8,
+				Jitter:   0.4,
+			},
+			func() (done bool, err error) {
+				if err := t.Get(ctx, client.ObjectKeyFromObject(o), oCopy); err != nil {
+					if apierrors.IsNotFound(err) {
+						return true, nil
+					}
+					return false, err
+				}
+				return false, nil
+			})
+		errs = append(errs, errors.Wrapf(err, "object %s, %s is not being deleted from the testenv client cache", o.GetObjectKind().GroupVersionKind().String(), client.ObjectKeyFromObject(o)))
+	}
+	return kerrors.NewAggregate(errs)
+}
+
 // CreateObj wraps around client.Create and creates the object.
+// Deprecated: use t.Client.Create or CreateAndWait.
 func (t *TestEnvironment) CreateObj(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
```

Comment:
Is this wrapper needed? Could we remove it?

Reply:
I there are no deprecation, yesssss